room.go

```diff
@@ -266,9 +288,31 @@ func (r *Room) JoinWithToken(url, token string, opts ...ConnectOption) error {
 		opt(params)
 	}
 
-	joinRes, err := r.engine.Join(url, token, params)
-	if err != nil {
-		return err
+	var joinRes *livekit.JoinResponse
+	cloudHostname, _ := parseCloudURL(url)
+	if params.DisableRegionDiscovery || cloudHostname == "" {
+		var err error
+		joinRes, err = r.engine.Join(url, token, params)
+		if err != nil {
+			return err
+		}
+	} else {
+		for joinRes == nil {
+			bestURL, err := r.regionURLProvider.BestURL(cloudHostname, token)
+			if err != nil {
+				return err
+			}
+
+			logger.Debugw("RTC engine joining room",
+				"url", bestURL,
+			)
+			joinRes, err = r.engine.Join(bestURL, token, params)
+			if err != nil {
+				// try the next URL
+				r.regionURLProvider.ReportAttemptFailure(cloudHostname, bestURL)
+				continue
```

Comment:
I think we do want to retry with the next best cluster/region without a delay. the JS SDK looks for either a 401 or a timeout before a retry, instead of always retrying. I wonder if that makes more sense, like this:
```
if err != nil {
  if os.IsTimeout(err) || strings.HasPrefix(err.Error(), "unauthorized: ") {
    // try the next URL
    r.regionURLProvider.ReportAttemptFailure(cloudHostname, bestURL)
    continue
  }
  return err
}
```